### PR TITLE
Static linking

### PR DESCRIPTION
**Pull request by chfast**
Now solc does not have any runtime deps besides C++ standard library. You can send it to your friend.

Also includes big pack of CMake cleanups.

### Before:
```
572K	solc/solc
1,2M	libevmasm/libsolevmasm.so
3,5M	libsolidity/libsolidity.so
336K	libdevcore/libsoldevcore.so
5,6M	total
```
### After:
```
5,5M       solc/solc
```

### CODE REVIEW

File: libevmasm/CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,6 @@
...
-file(GLOB HEADERS "*.h")
-
-include_directories(BEFORE ..)
-add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-eth_use(${EXECUTABLE} REQUIRED Dev::soldevcore)
-target_link_libraries(${EXECUTABLE} jsoncpp)
-
-install( TARGETS ${EXECUTABLE} RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib )
+add_library(evmasm ${sources} ${headers})
+target_link_libraries(evmasm PUBLIC devcore jsoncpp)
```
Comment (chriseth) [inline libevmasm/CMakeLists.txt:5]: Why is jsoncpp public here but boost is private in libdevcore?

File: .travis.yml [hunk 2]
```diff
@@ -174,7 +174,6 @@ cache:
     ccache: true
     directories:
         - boost_1_57_0
-        - build
```
Comment (chriseth) [inline .travis.yml]: I that just a temporary change?

Comment (chfast) [reply .travis.yml]: I hope not :) Build without CMake cache are always better.

File: libevmasm/CMakeLists.txt (hunk 1 above)
Comment (chfast) [reply libevmasm/CMakeLists.txt:5]: This is more/less guess and try. I'm trying to find the minimal set. We don't have nice library structure here. jsoncpp must be in PUBLIC, because headers of devcore includes jsoncpp headers.

File: .travis.yml (hunk 2 above)
Comment (chriseth) [reply .travis.yml]: Ah you mean it does not include the build directory but it does include the ccache directory?

Comment (chriseth) [reply .travis.yml]: If that is the case then it's fine.

**Comment by chfast** [comment]
Finally! This is now only refactoring and cleanups. I moved CMake toolchain support to other PR.

**Comment by pirapira** [comment]
Many lines disappeared nicely.  What kind of things shall I be careful about when reviewing cmake PRs?

**Comment by chfast** [comment]
@pirapira Nothing more than checking if the scripts are as simple as possible and easy to read.

Does PPA builds expect shared libraries to be built?

**Comment by axic** [comment]
@chfast can de emscripten cleanup (last commit) merged separately? It seems so.

**Comment by chfast** [comment]
@axic yes. Actually I have more changes there to send...

**Comment by chfast** [comment]
@axic Done. CI is building it...

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -1,26 +1,9 @@
...
-file(GLOB HEADERS "*.h" "*/*.h")
-set(EXECUTABLE soltest)
-eth_simple_add_executable(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-
-eth_use(${EXECUTABLE} REQUIRED Solidity::solidity Solidity::lll)
-
-include_directories(BEFORE ..)
-target_link_libraries(${EXECUTABLE} soljson ${Boost_UNIT_TEST_FRAMEWORK_LIBRARIES})
+add_executable(soltest ${sources} ${headers})
+target_link_libraries(soltest PRIVATE soljson solidity lll ${Boost_UNIT_TEST_FRAMEWORK_LIBRARIES})
```
Comment (axic) [inline test/CMakeLists.txt]: This depends on `evmasm` too.

Comment (axic) [reply test/CMakeLists.txt]: Also both depend on devcore if you want to make explicit.

File: libdevcore/IndentedWriter.cpp [hunk 4]
```diff
@@ -27,6 +27,7 @@ using namespace dev;
 
 string IndentedWriter::format() const
 {
+
```
Comment (axic) [inline libdevcore/IndentedWriter.cpp]: Unrelated.

**Comment by axic** [comment]
Do you want to pull out the "remove unused scripts" I think that can be merged without much review.

I see that the `EthExecutableHelper` is used from there before the other changes.

File: test/CMakeLists.txt (hunk 3 above)
Comment (chfast) [reply test/CMakeLists.txt]: solidity depends on evmasm so this is transitional. Does soltest include anything from evmasm directly?

Comment (chfast) [reply test/CMakeLists.txt]: I'd keep it as long as it builds fine.

**Comment by chfast** [comment]
How about now?

File: test/CMakeLists.txt (hunk 3 above)
Comment (axic) [reply test/CMakeLists.txt]: It does have direct `evmasm` tests.

File: cmake/EthCompilerSettings.cmake [hunk 5]
```diff
@@ -63,13 +63,6 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 	# Applying -fpermissive to a C command-line (ie. secp256k1) gives a build error.
 	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fpermissive")
 
-	# Build everything as shared libraries (.so files)
-	add_definitions(-DSHAREDLIB)
-
-	# If supported for the target machine, emit position-independent code, suitable for dynamic
-	# linking and avoiding any limit on the size of the global offset table.
-	add_compile_options(-fPIC)
```
Comment (axic) [inline cmake/EthCompilerSettings.cmake:71]: Is `PIC` added back by cmake if `BUILD_SHARED_LIBS` is on?

File: cmake/EthDependencies.cmake [hunk 6]
```diff
@@ -43,62 +43,9 @@ find_program(CTEST_COMMAND ctest)
...
-		set(CMAKE_FIND_LIBRARY_SUFFIXES .a .dylib)
-	else()
-		set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
-	endif()
-
-	set(ETH_STATIC ON)
-endif()
-
-find_package(Boost 1.54.0 QUIET REQUIRED COMPONENTS thread date_time system regex chrono filesystem unit_test_framework program_options random)
+find_package(Boost 1.54.0 QUIET REQUIRED COMPONENTS regex filesystem unit_test_framework program_options)
```
Comment (axic) [inline cmake/EthDependencies.cmake:48]: `chrono` is used in tests.

File: liblll/CMakeLists.txt [hunk 7]
```diff
@@ -1,18 +1,6 @@
...
-
-include_directories(BEFORE ..)
-add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-
-eth_use(${EXECUTABLE} REQUIRED Solidity::solevmasm)
-#target_link_libraries(${EXECUTABLE} evmasm)
-
-install( TARGETS ${EXECUTABLE} RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib )
+add_library(lll ${sources} ${headers})
+target_link_libraries(lll PUBLIC evmasm)
```
Comment (axic) [inline liblll/CMakeLists.txt]: I think this also uses `devcore` directly (not only via `evmasm`).

File: test/CMakeLists.txt (hunk 3 above)
Comment (axic) [reply test/CMakeLists.txt]: Also it has direct tests with `devcore`.

File: cmake/EthCompilerSettings.cmake (hunk 5 above)
Comment (chfast) [reply cmake/EthCompilerSettings.cmake:71]: Yes.

File: cmake/EthDependencies.cmake (hunk 6 above)
Comment (chfast) [reply cmake/EthDependencies.cmake:48]: Only `std::chrono` is used.

Comment (axic) [reply cmake/EthDependencies.cmake:48]: Oh, that's correct.

Comment (axic) [reply cmake/EthDependencies.cmake:48]: And `regex` is actually not used IIRC as the construct we wanted to use was not available in old Boost versions.

File: liblll/CMakeLists.txt (hunk 7 above)
Comment (chfast) [reply liblll/CMakeLists.txt]: I made devcore dep explicit.

File: cmake/EthDependencies.cmake (hunk 6 above)
Comment (chfast) [reply cmake/EthDependencies.cmake:48]: https://github.com/ethereum/solidity/blob/develop/libdevcore/Whiskers.cpp#L28

Comment (axic) [reply cmake/EthDependencies.cmake:48]: Ok, so that means in the other cases we tried using (see `test/JSONCompiler.cpp` which still has the include) regex from the std library.

File: libdevcore/CMakeLists.txt [hunk 8]
```diff
@@ -1,14 +1,8 @@
...
-add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-
-eth_use(${EXECUTABLE} REQUIRED Dev::base)
-
-install( TARGETS ${EXECUTABLE} RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib )
+file(GLOB sources "*.cpp")
+file(GLOB headers "*.h")
+
+add_library(devcore ${sources} ${headers})
+target_link_libraries(devcore PRIVATE ${Boost_FILESYSTEM_LIBRARIES} ${Boost_REGEX_LIBRARIES} ${Boost_SYSTEM_LIBRARIES})
```
Comment (axic) [inline libdevcore/CMakeLists.txt]: `Boost_SYSTEM_LIBRARIES` doesn't seem to be used here

File: libsolidity/CMakeLists.txt [hunk 9]
```diff
@@ -1,23 +1,7 @@
...
-add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-
-eth_use(${EXECUTABLE} REQUIRED Dev::soldevcore Solidity::solevmasm)
-
-install( TARGETS ${EXECUTABLE} RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib )
+file(GLOB_RECURSE sources "*.cpp" "../libjulia/*.cpp")
+file(GLOB_RECURSE headers "*.h" "../libjulia/*.h")
 
+add_library(solidity ${sources} ${headers})
+target_link_libraries(solidity PUBLIC evmasm devcore PRIVATE ${Boost_REGEX_LIBRARIES})
```
Comment (axic) [inline libsolidity/CMakeLists.txt]: Regex doesn't seem to be used in here.

File: libdevcore/CMakeLists.txt (hunk 8 above)
Comment (chfast) [reply libdevcore/CMakeLists.txt]: Not directly, but boost::filesystem depends on it. It will go away if we use imported libs provided from FindBoost CMake module from version 3.5+.
https://cmake.org/cmake/help/v3.5/module/FindBoost.html

File: cmake/EthDependencies.cmake [hunk 10]
```diff
@@ -43,62 +43,9 @@ find_program(CTEST_COMMAND ctest)
 
 ## use multithreaded boost libraries, with -mt suffix
 set(Boost_USE_MULTITHREADED ON)
+set(Boost_USE_STATIC_LIBS ON)
```
Comment (chriseth) [inline cmake/EthDependencies.cmake]: Why is this on unconditionally?

File: libsolidity/CMakeLists.txt [hunk 11]
```diff
@@ -1,23 +1,7 @@
...
-
-file(GLOB HEADERS "*/*.h" "../libjulia/backends/evm/*")
-
-include_directories(BEFORE ..)
-add_library(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
-
-eth_use(${EXECUTABLE} REQUIRED Dev::soldevcore Solidity::solevmasm)
-
-install( TARGETS ${EXECUTABLE} RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib )
+file(GLOB_RECURSE sources "*.cpp" "../libjulia/*.cpp")
```
Comment (chriseth) [inline libsolidity/CMakeLists.txt:2]: This weird, because the cpp files are actually in subdirectories of `libjulia`, but if it still works...

**Review by chriseth — APPROVED** [review]

File: libsolidity/CMakeLists.txt (hunk 11 above)
Comment (chfast) [reply libsolidity/CMakeLists.txt:2]: See: `GLOB_RECURSE`.

File: cmake/EthDependencies.cmake (hunk 10 above)
Comment (chfast) [reply cmake/EthDependencies.cmake]: To link statically also on Linux. I can check what the difference is in size.

Comment (chriseth) [reply cmake/EthDependencies.cmake]: I think if we build for a package manager, we should not link statically.

Comment (chfast) [reply cmake/EthDependencies.cmake]: Before:
libsolidity.so: 52M
libsoldevcore.so: 3M
libsolevmasm.so: 17M
solc: 5M
total: 77M

After:
solc: 66M

Comment (chriseth) [reply cmake/EthDependencies.cmake]: Please only measure the release binary. If the release binary is 66M, we are in serious trouble.

**Comment by chriseth** [comment]
Ok, great, this now produces a correctly statically linked binary!
@axic please merge if you like it.

File: docs/installing-solidity.rst [hunk 12]
```diff
@@ -264,6 +265,11 @@ Alternatively, you can build for Windows on the command-line, like so:
 
     cmake --build . --config RelWithDebInfo
 
+CMake options
+=============
+
+If you are interested what CMake options are available run ``cmake .. -LH``.
```
Comment (axic) [inline docs/installing-solidity.rst:271]: Is there a way to make this also include `TESTS`, `PROFILING`, etc.?

Comment (chfast) [reply docs/installing-solidity.rst:271]: Yes. They must be "cached variables". I will fix it next time.

File: solc/CMakeLists.txt [hunk 13]
```diff
@@ -1,27 +1,28 @@
...
-	eth_use(soljson REQUIRED Solidity::solidity)
+	add_executable(soljson jsonCompiler.cpp)
 else()
-	add_library(soljson jsonCompiler.cpp ${HEADERS})
-	target_link_libraries(soljson solidity)
+	add_library(soljson jsonCompiler.cpp)
 endif()
+target_link_libraries(soljson PRIVATE solidity)
+
+if(SOLC_LINK_STATIC AND UNIX AND NOT APPLE)
```
Comment (axic) [inline solc/CMakeLists.txt:21]: Why is this disabled on Mac?

Comment (chfast) [reply solc/CMakeLists.txt:21]: I don't know. It was like this before. I'm guessing linker on macOS does not support this or it is enabled by default.

File: solc/CMakeLists.txt [hunk 14]
```diff
@@ -16,4 +16,16 @@ if (EMSCRIPTEN)
 else()
 	add_library(soljson jsonCompiler.cpp)
 endif()
-target_link_libraries(soljson PRIVATE solidity)
\ No newline at end of file
+target_link_libraries(soljson PRIVATE solidity)
+
+if(SOLC_LINK_STATIC AND UNIX AND NOT APPLE)
+	# Produce solc as statically linked binary (includes C/C++ standard libraries)
+	# This is supported on macOS, see
```
Comment (axic) [inline solc/CMakeLists.txt]: unsupported :)

**Comment by chfast** [comment]
@axic approved?
